docs/dev/development/patterns.rst

```diff
@@ -107,6 +107,107 @@ Class                     Method
 
+Implementing new services
+=========================
+
+Warehouse uses services to provide pluggable functionalities within the codebase. They are implemented using
+`pyramid-service`_. After being registered, services are accessible using the ``find_service`` method of the
+``request`` object.
+
+When adding new services to ``warehouse``, the following checklist serves as a comprehensive guideline to ensure
+you stay on track.
+
+Adding a new service
+~~~~~~~~~~~~~~~~~~~~~
+
+1. Create an Interface for the service. The interface serves as the baseline of the new service (design by
+   contract pattern) and details all methods and attributes shared by the different service implementations.
+
+   Warehouse uses zope_ to define interfaces. The interfaces are usually declared in a file named
+   ``interfaces.py`` within the relevant component, such as ``packaging/interfaces.py``.
+
+2. Create the new service. The service must define all methods and attributes declared in the interface.
+   This implementation contains the core logic of the service features.
+
+3. (Optional) Create other implementations of the interface. For instance, many services in ``warehouse``
+   also provide a ``NullService`` version used for development. These Null implementations only
+   provide basic functionalities without verifications and reduce the need for stubs in tests.
+
+   Any new implementation must implement the complete interface, including all its methods and attributes.
+
+4. Implement each service creation method. If the Service is simple enough, use a class method in
+   your service implementation (usually named ``create_service``). For more complex cases, implement
+   a ``ServiceFactory`` class, responsible to create the service instance.
+
+5. Register the service. The new service(s) must be registered to be available in the request object.
+
+   - If you have multiple services, create a new setting (in ``warehouse/config.py``) to select
+     which backend to use.
+
+   - Add a default value for the setting in ``dev/environment`` for the development environment.
+
+   - Use the setting value in the ``includeme`` function to instantiate the appropriate service.
+
+   - Register your service factory. This registration must be in the service module's ``includeme``
+     function for Pyramid to detect it and use the service factory created at the previous step.
+
+6. (Optional) Add the new module to the ``warehouse/config.py``. If the new service is defined in a
+   new module, add the new module within the warehouse ``configure`` function. This enrollment
+   ensures Pyramid can detect it.
+
+Using the service
+~~~~~~~~~~~~~~~~~
+
+To use a service, query it using ``request.find_services`` with the service interface. This
+method will return an instance of the service correctly selected based on the context and environment.
+
+Example:
+
+.. code-block:: python
+
+   metrics = request.find_service(IMetricsService, context=None)
```

Comment:
thought: `context=` is not always required/used, nor is `name=` - would it be better to show an example without `context=None`, or have examples for using named service implementations?